Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h

```diff
@@ -40,11 +41,16 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
 private:
     explicit WebStorageNamespaceProvider(StorageNamespaceIdentifier localStorageIdentifier);
 
-    Ref<WebCore::StorageNamespace> createSessionStorageNamespace(WebCore::Page&, unsigned quota) override;
+    Ref<WebCore::StorageNamespace> createSessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, unsigned quota) override;
     Ref<WebCore::StorageNamespace> createLocalStorageNamespace(unsigned quota, PAL::SessionID) override;
     Ref<WebCore::StorageNamespace> createTransientLocalStorageNamespace(WebCore::SecurityOrigin&, unsigned quota, PAL::SessionID) override;
 
+    RefPtr<WebCore::StorageNamespace> sessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, ShouldCreateIfNotExists) override;
+    void setSessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, RefPtr<WebCore::StorageNamespace>&&) override;
+
     const StorageNamespaceIdentifier m_localStorageNamespaceIdentifier;
+
+    HashMap<std::pair<StorageNamespaceImpl::Identifier, WebCore::SecurityOriginData>, RefPtr<WebCore::StorageNamespace>> m_sessionStorageNamespaces;
```

Comment:
WebPage's WebStorageNamespaceProvider is instantiated in WebPage's constructor as:
`     pageConfiguration.storageNamespaceProvider = WebStorageNamespaceProvider::getOrCreate(*m_pageGroup);`
That resulted in my understanding that the StorageNamespaceProvider is shared across pages within a Page Group (keyed by the PageGroup's pageGroupID). Therefore, I believe the SessionStorageNamespaces map is only isolated by PageGroup, and not per WebPage. I'll make the SessionStorageNamespaces map static in WebStorageNamespaceProvider.cpp, as you suggest, though. I believe we still need both WebPageProxyIdentifier and top-level origin as keys, if my understanding is correct.

Reply:
ah, you are right! Actually we have one StorageNamespaceProvider per PageGroup, and I think by default pages are in the same group. So your change is actually fine; you don't need to move m_sessionStorageNamespaces!
But since the WebStorageNamespaceProvider is shared across multiple WebPages, I think we still need a way to destroy StorageNameSpace when WebPage is gone, otherwise there can be abandoned memory (i.e. WebPageProxy is gone but we still keep its SessionStorageNamespace in web process)

What I mean by using Identifier as the key is using Identifier as first-level key to find SessionStorageNamespaces: HashMap<Identifier, SessionStorageNamespaces> and struct SessionStorageNamespaces keeps track of use count (WebPage count, which is updated when WebPage is created and destroyed) and HashMap<SecurityOriginData, StorageNameSpace>. 
This is just an idea of how to get rid of SessionStorageNamespaces that will not be used.

Reply:
Yes, that's a good idea. My current understanding is that the existing implementation has the same problem, and that is why I didn't worry about it very much (but I'm happy to address it now).